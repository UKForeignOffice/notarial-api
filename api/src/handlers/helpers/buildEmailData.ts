import { getAllInputsFromForm } from "./getAllInputsFromForm";
import { FormDataBody } from "../../types";
import { getTemplateDataFromInputs } from "./getTemplateDataFromInputs";
import { HttpException } from "../../middlewares/error-handlers";
import { ERRORS } from "../../errors";

export type Errors = {
  errors: Error;
};

export function buildEmailData(
  formBody: FormDataBody,
  formType: "cni" | "affirmation"
) {
  const fields = getAllInputsFromForm(formBody);
  if (!fields) {
    throw new HttpException(400, "400", ERRORS.webhook.EMPTY_TEMPLATE_DATA);
  }
  const templateData = getTemplateDataFromInputs(fields, formType);
  return templateData;
}
