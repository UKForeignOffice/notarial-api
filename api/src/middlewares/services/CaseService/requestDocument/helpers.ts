import { order, remap } from "./mappings";
import { createRemapper } from "../utils/createRemapper";
import { reorderSectionsWithNewName } from "../utils/reorderSectionsWithNewName";

export const requestDocumentRemapper = createRemapper(remap);
export const requestDocumentReorderer = reorderSectionsWithNewName(order);
