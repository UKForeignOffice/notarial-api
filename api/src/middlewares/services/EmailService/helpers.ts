import { FormField } from "../../../types/FormField";

export function fieldsHashMap(fields: FormField[]): Record<string, FormField> {
  return fields.reduce(
    (map, currField) => ({
      ...map,
      [currField.key]: currField,
    }),
    {}
  );
}

export function getFileFields(fields: Record<string, FormField>) {
  return Object.values(fields).reduce<FormField[]>((acc, field) => {
    if (field.type === "FileUploadField") {
      return [...acc, field];
    }
    return [...acc];
  }, []);
}
