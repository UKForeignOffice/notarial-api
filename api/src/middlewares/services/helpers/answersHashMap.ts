import { FormField } from "../../../types/FormField";
import { AnswersHashMap } from "../../../types/AnswersHashMap";

export function answersHashMap(fields: FormField[]) {
  return fields.reduce<AnswersHashMap>(
    (acc, field) => ({
      ...acc,
      [field.key]: field.answer as string | boolean,
    }),
    {}
  );
}
