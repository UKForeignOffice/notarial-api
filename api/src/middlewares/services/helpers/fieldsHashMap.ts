import { FormField } from "../../../types/FormField";

export function fieldsHashMap(fields: FormField[]): Record<string, FormField> {
  return fields.reduce(
    (map, currField) => ({
      ...map,
      [currField.key]: currField,
    }),
    {}
  );
}
