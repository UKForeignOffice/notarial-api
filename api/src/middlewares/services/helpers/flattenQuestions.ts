import { FormQuestion } from "../../../types/FormQuestion";

export function flattenQuestions(questions: FormQuestion[]) {
  return questions.flatMap(({ fields }) => {
    return fields;
  });
}
