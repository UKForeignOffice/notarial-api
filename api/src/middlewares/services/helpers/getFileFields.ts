import { FormField } from "../../../types/FormField";

export function getFileFields(fields: FormField[]) {
  return fields.reduce<FormField[]>((acc, field) => {
    if (field.type === "file") {
      return [...acc, field];
    }
    return [...acc];
  }, []);
}
