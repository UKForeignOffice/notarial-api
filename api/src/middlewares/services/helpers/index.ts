export { flattenQuestions } from "./flattenQuestions";
export { answersHashMap } from "./answersHashMap";
export { getFileFields } from "./getFileFields";
