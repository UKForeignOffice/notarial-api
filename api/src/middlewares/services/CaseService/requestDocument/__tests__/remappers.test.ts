import { flattenQuestions } from "../../../helpers";
import { requestDocumentRemapper, requestDocumentReorderer } from "../helpers";
import { data } from "./fixtures/data";

test("requestDocument is successfully remapped", () => {
  const flattened = flattenQuestions(data.data.questions);
  const remapped = requestDocumentRemapper(flattened);
  console.log(remapped);
  expect(remapped).toMatchSnapshot();
});

test("requestDocument is successfully reordered", () => {
  const flattened = flattenQuestions(data.data.questions);
  const remapped = requestDocumentRemapper(flattened);
  const reordered = requestDocumentReorderer(remapped);
  console.log(reordered);
  expect(reordered).toMatchSnapshot();
});
