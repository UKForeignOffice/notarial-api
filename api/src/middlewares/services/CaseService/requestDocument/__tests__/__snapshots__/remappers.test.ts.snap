// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requestDocument is successfully remapped 1`] = `
{
  "applicantDetails": {
    "firstName": {
      "answer": "George",
      "category": "applicantDetails",
      "key": "firstName",
      "title": "First name",
      "type": "text",
    },
    "middleName": {
      "answer": "G",
      "category": "applicantDetails",
      "key": "middleName",
      "title": "Middle name",
      "type": "text",
    },
    "surname": {
      "answer": "Goofy",
      "category": "applicantDetails",
      "key": "surname",
      "title": "Surname",
      "type": "text",
    },
  },
  "applicantPassport": {
    "countryOfBirth": {
      "answer": "United Kingdom",
      "category": "applicantDetails",
      "key": "countryOfBirth",
      "title": "Country of birth",
      "type": "list",
    },
    "dateOfBirth": {
      "answer": "2000-12-25",
      "category": "applicantDetails",
      "key": "dateOfBirth",
      "title": "Date of birth",
      "type": "date",
    },
    "dateOfIssue": {
      "answer": "2000-12-25",
      "category": "applicantDetails",
      "key": "passportDateOfIssue",
      "title": "Date of issue of passport",
      "type": "date",
    },
    "number": {
      "answer": "123123",
      "category": "applicantDetails",
      "key": "passportNumber",
      "title": "Passport number",
      "type": "text",
    },
    "placeOfBirth": {
      "answer": "london",
      "category": "applicantDetails",
      "key": "placeOfBirth",
      "title": "Place of birth ",
      "type": "text",
    },
  },
  "applicationForm": {
    "answer": "http://documentupload:9000/v1/files/a",
    "category": "documents",
    "key": "applicationForm",
    "title": "Upload your application form",
    "type": "file",
  },
  "contactDetails": {
    "emailAddress": {
      "answer": "test@test.com",
      "category": "contact",
      "key": "emailAddress",
      "title": "Confirm email address",
      "type": "text",
    },
    "phoneNumber": {
      "answer": "+1 23",
      "category": "contact",
      "key": "phoneNumber",
      "title": "Confirm phone number",
      "type": "text",
    },
  },
  "declaration": {
    "answer": true,
    "category": null,
    "key": "declaration",
    "title": "Declaration",
    "type": "boolean",
  },
  "feedbackConsent": {
    "feedbackConsent": {
      "answer": false,
      "category": "feedback",
      "key": "feedbackConsent",
      "title": "Can our partner contact you for feedback to help improve this service?",
      "type": "list",
    },
  },
  "serviceType": {
    "answer": "USA - J1 visa no objection statement",
    "category": "beforeYouStart",
    "key": "serviceType",
    "title": "Service being applied for",
    "type": "list",
  },
  "thirdPartyBarcode": {
    "answer": "http://documentupload:9000/v1/files/b",
    "category": "documents",
    "key": "thirdPartyBarcode",
    "title": "Upload your third party barcode page",
    "type": "file",
  },
  "ukPassportFile": {
    "answer": "http://documentupload:9000/v1/files/e95a22c4-6dd7-41b1-98cb-61685f508969.jpg",
    "category": "documents",
    "key": "ukPassportFile",
    "title": "UK passport",
    "type": "file",
  },
}
`;

exports[`requestDocument is successfully reordered 1`] = `
{
  "Applicant's details": {
    "firstName": {
      "answer": "George",
      "category": "applicantDetails",
      "key": "firstName",
      "title": "First name",
      "type": "text",
    },
    "middleName": {
      "answer": "G",
      "category": "applicantDetails",
      "key": "middleName",
      "title": "Middle name",
      "type": "text",
    },
    "surname": {
      "answer": "Goofy",
      "category": "applicantDetails",
      "key": "surname",
      "title": "Surname",
      "type": "text",
    },
  },
  "Applicant's passport details": {
    "countryOfBirth": {
      "answer": "United Kingdom",
      "category": "applicantDetails",
      "key": "countryOfBirth",
      "title": "Country of birth",
      "type": "list",
    },
    "dateOfBirth": {
      "answer": "2000-12-25",
      "category": "applicantDetails",
      "key": "dateOfBirth",
      "title": "Date of birth",
      "type": "date",
    },
    "dateOfIssue": {
      "answer": "2000-12-25",
      "category": "applicantDetails",
      "key": "passportDateOfIssue",
      "title": "Date of issue of passport",
      "type": "date",
    },
    "number": {
      "answer": "123123",
      "category": "applicantDetails",
      "key": "passportNumber",
      "title": "Passport number",
      "type": "text",
    },
    "placeOfBirth": {
      "answer": "london",
      "category": "applicantDetails",
      "key": "placeOfBirth",
      "title": "Place of birth ",
      "type": "text",
    },
  },
  "Certificate details": undefined,
  "Contact details": {
    "emailAddress": {
      "answer": "test@test.com",
      "category": "contact",
      "key": "emailAddress",
      "title": "Confirm email address",
      "type": "text",
    },
    "phoneNumber": {
      "answer": "+1 23",
      "category": "contact",
      "key": "phoneNumber",
      "title": "Confirm phone number",
      "type": "text",
    },
  },
  "Delivery address": undefined,
  "Feedback consent": {
    "feedbackConsent": {
      "answer": false,
      "category": "feedback",
      "key": "feedbackConsent",
      "title": "Can our partner contact you for feedback to help improve this service?",
      "type": "list",
    },
  },
  "Marital status": undefined,
  "Marriage details": undefined,
  "Partner's details": undefined,
  "Reason for document": undefined,
}
`;
